packages/core/src/dispatcher.ts

```diff
@@ -128,23 +121,49 @@ export class Dispatcher {
    */
   async retrieveFromIPFS(cid: CID | string, path?: string): Promise<any> {
     try {
-      const asCid = typeof cid === 'string' ? new CID(cid) : cid
-
-      // Lookup CID in cache before looking it up IPFS
-      const cidAndPath = path ? asCid.toString() + path : asCid.toString()
-      const cachedDagNode = await this.dagNodeCache.get(cidAndPath)
-      if (cachedDagNode) return cloneDeep(cachedDagNode)
-
-      // Now lookup CID in IPFS and also store it in the cache
-      const dagNode = await this._ipfs.dag.get(asCid, { timeout: IPFS_GET_TIMEOUT, path })
-      await this.dagNodeCache.set(cidAndPath, dagNode.value)
-      return cloneDeep(dagNode.value)
+      return this._getFromIpfs(cid, path)
     } catch (e) {
       this._logger.err(`Error while loading CID ${cid.toString()} from IPFS: ${e}`)
       throw e
     }
   }
 
+  /**
+   * Helper function for loading a CID from IPFS
+   */
+  private async _getFromIpfs(cid: CID | string, path?: string): Promise<any> {
+    const asCid = typeof cid === 'string' ? new CID(cid) : cid
+
+    // Lookup CID in cache before looking it up IPFS
+    const cidAndPath = path ? asCid.toString() + path : asCid.toString()
+    const cachedDagNode = await this.dagNodeCache.get(cidAndPath)
+    if (cachedDagNode) return cloneDeep(cachedDagNode)
+
+    // Now lookup CID in IPFS, with retry logic
+    // Note, in theory retries shouldn't be necessary, as just increasing the timeout should
+    // allow IPFS to use the extra time to find the CID, doing internal retries if needed.
+    // Anecdotally, however, we've seen evidence that IPFS sometimes finds CIDs on retry that it
+    // doesn't on the first attempt, even when given plenty of time to load it.
+    let dagResult = null
+    for (let retries = IPFS_GET_RETRIES - 1; retries >= 0 && dagResult == null; retries--) {
+      try {
+        dagResult = await this._ipfs.dag.get(asCid, {timeout: IPFS_GET_TIMEOUT, path})
```

Comment:
It's generally considered good practice to use exponential backoff with jitter when retrying.

The reason for the exponential backoff is so that a struggling network/system doesn't get slammed in tight succession with retries. The reason for the jitter is to make sure that retries from different parts of the system don't all get aligned at a single point in time and cause the network/system to get slammed when the backoff kicks in.

One or both of these are optional but would be good to consider if they would be useful to add. Neither will make the logic significantly more complicated.

Reply:
with the timeout being so high (30 seconds) I'm not too worried about hammering the IPFS node with these retries.

I'd rather keep things simple for now and add more advanced retry logic here if/when we see it being a problem